cmd/xl-storage-format-v2.go

```diff
@@ -0,0 +1,437 @@
+		}
+		jentry.Object.DataPartInfoNumbers[i] = fi.Parts[i].Number
+		jentry.Object.DataPartInfoActualSizes[i] = fi.Parts[i].ActualSize
+	}
+
+	for k, v := range fi.Metadata {
+		if strings.HasPrefix(k, ReservedMetadataPrefix) || strings.HasPrefix(k, strings.ToLower(ReservedMetadataPrefix)) {
+			jentry.Object.MetaSys[k] = []byte(v)
+		} else {
+			jentry.Object.MetaUser[k] = v
+		}
+	}
+	z.ObjectJournals = append(z.ObjectJournals, jentry)
+	return nil
+}
+
+func newXLMetaV2(fi FileInfo) (XLMetaV2, error) {
+	xlMeta := XLMetaV2{}
+
+	var uv uuid.UUID
+	var err error
+	if fi.VersionID != nullVersionID && fi.VersionID != "" {
+		uv, err = uuid.Parse(fi.VersionID)
+		if err != nil {
+			return xlMeta, err
+		}
+	}
+	dd, err := uuid.Parse(fi.DataDir)
+	if err != nil {
+		return xlMeta, err
+	}
+	jentry := XLMetaV2JournalEntry{
+		Type: ObjectType,
+		Object: &XLMetaV2Object{
+			VersionID:               uv,
+			DataDir:                 dd,
+			StatSize:                fi.Size,
+			StatModTime:             fi.ModTime.Unix(),
+			DataErasureAlgorithm:    ReedSolomon,
+			DataErasureM:            fi.Erasure.DataBlocks,
+			DataErasureN:            fi.Erasure.ParityBlocks,
+			DataErasureBlockSize:    fi.Erasure.BlockSize,
+			DataErasureIndex:        fi.Erasure.Index,
+			DataErasureChecksumAlgo: HighwayHash,
+			DataErasureDistribution: make([]uint8, len(fi.Erasure.Distribution)),
+			DataPartInfoNumbers:     make([]int, len(fi.Parts)),
+			DataPartInfoETags:       make([]string, len(fi.Parts)),
+			DataPartInfoSizes:       make([]int64, len(fi.Parts)),
+			DataPartInfoActualSizes: make([]int64, len(fi.Parts)),
+			MetaSys:                 make(map[string][]byte),
+			MetaUser:                make(map[string]string),
+		},
+	}
+
+	for i := range fi.Erasure.Distribution {
+		jentry.Object.DataErasureDistribution[i] = uint8(fi.Erasure.Distribution[i])
+	}
+
+	for i := range fi.Parts {
+		jentry.Object.DataPartInfoSizes[i] = fi.Parts[i].Size
+		if fi.Parts[i].ETag != "" {
+			jentry.Object.DataPartInfoETags[i] = fi.Parts[i].ETag
+		}
+		jentry.Object.DataPartInfoNumbers[i] = fi.Parts[i].Number
+		jentry.Object.DataPartInfoActualSizes[i] = fi.Parts[i].ActualSize
+	}
+
+	for k, v := range fi.Metadata {
+		if strings.HasPrefix(k, ReservedMetadataPrefix) || strings.HasPrefix(k, strings.ToLower(ReservedMetadataPrefix)) {
+			jentry.Object.MetaSys[k] = []byte(v)
+		} else {
+			jentry.Object.MetaUser[k] = v
+		}
+	}
+
+	xlMeta.ObjectJournals = []XLMetaV2JournalEntry{jentry}
+	return xlMeta, nil
+}
+
+var (
+	nullVersionIDBytes = uuid.UUID{}
+)
+
+// ToFileInfo - convert a object journal entry to file info.
+func (j XLMetaV2JournalEntry) ToFileInfo(volume, path string) (FileInfo, error) {
+	fi := FileInfo{
+		Volume:    volume,
+		Name:      path,
+		VersionID: nullVersionID,
+		Metadata:  make(map[string]string),
+	}
+
+	var oj XLMetaV2Object
+	switch j.Type {
+	case ObjectType:
+		oj = *j.Object
+	case DeleteType:
+		return FileInfo{}, errFileNotFound
+	}
```

Comment:
This copies the entire XLMetaV2Object struct and it is only used in this case,

```suggestion
	switch j.Type {
	case ObjectType:
	case DeleteType:
		return FileInfo{}, errFileNotFound
	default:
		return FileInfo{}, fmt.Errorf("unknown journal type: %v", j.Type)
	}

	oj := j.Object
```

Maybe sprinkle a nil check of j.Object in there as well.